materialize-starburst-galaxy/galaxy.go

```diff
@@ -0,0 +1,405 @@
+	return c.withDB(func(db *stdsql.DB) error { return sql.StdSQLExecStatements(ctx, db, statements) })
+}
+
+func (c client) InstallFence(_ context.Context, _ sql.Table, _ sql.Fence) (sql.Fence, error) {
+	return sql.Fence{}, nil
+}
+
+func (c client) openDB() (*stdsql.DB, error) {
+	return stdsql.Open("trino", c.uri)
+}
+
+func (c client) connectToDb(ctx context.Context) (*stdsql.Conn, error) {
+	db, err := c.openDB()
+	if err != nil {
+		return nil, fmt.Errorf("stdsql.Open failed: %w", err)
+	}
+	return db.Conn(ctx)
+}
+
+func (c client) withDB(fn func(*stdsql.DB) error) error {
+	db, err := c.openDB()
+	if err != nil {
+		return err
+	}
+	err = fn(db)
+	if err != nil {
+		db.Close()
+		return err
+	}
+	return db.Close()
+}
+
+type transactor struct {
+	cfg  *config
+	load struct {
+		conn *stdsql.Conn
+	}
+	store struct {
+		conn *stdsql.Conn
+	}
+	bindings []*binding
+}
+
+func newTransactor(
+	ctx context.Context,
+	ep *sql.Endpoint,
+	_ sql.Fence,
+	bindings []sql.Table,
+	_ pm.Request_Open,
+) (_ pm.Transactor, err error) {
+	var cfg = ep.Config.(*config)
+	var c = ep.Client.(client)
+	var templates = c.templates
+
+	var result = &transactor{
+		cfg: cfg,
+	}
+
+	// Establish connections.
+	result.load.conn, err = c.connectToDb(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("connection for load failed: %w", err)
+	}
+	result.store.conn, err = c.connectToDb(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("connection for store failed: %w", err)
+	}
+
+	for _, binding := range bindings {
+		if err = result.addBinding(binding, templates); err != nil {
+			return nil, fmt.Errorf("adding binding %v failed: %w", binding, err)
+		}
+	}
+	return result, nil
+}
+
+type binding struct {
+	target sql.Table
+	load   struct {
+		loadQuery string
+	}
+	store struct {
+		insert string
+		update string
+	}
+}
+
+func (t *transactor) addBinding(target sql.Table, templates templates) error {
+	var d = new(binding)
+	var err error
+	d.target = target
+
+	if d.store.insert, err = sql.RenderTableTemplate(target, templates.storeInsert); err != nil {
+		return fmt.Errorf("storeInsert template: %w", err)
+	}
+	if d.store.update, err = sql.RenderTableTemplate(target, templates.storeUpdate); err != nil {
+		return fmt.Errorf("storeUpdate template: %w", err)
+	}
+	if d.load.loadQuery, err = sql.RenderTableTemplate(target, templates.loadQuery); err != nil {
+		return fmt.Errorf("loadQuery template: %w", err)
+	}
+
+	t.bindings = append(t.bindings, d)
+	return nil
+}
+
+func (t *transactor) Load(it *pm.LoadIterator, loaded func(int, json.RawMessage) error) error {
+	it.WaitForAcknowledged()
+	for it.Next() {
+		var b = t.bindings[it.Binding]
+		converted, err := b.target.ConvertKey(it.Key)
+		if err != nil {
+			return fmt.Errorf("converting Load key: %w", err)
+		}
+		rows, err := t.load.conn.QueryContext(it.Context(), b.load.loadQuery, converted...)
+		if err != nil {
+			return fmt.Errorf("querying Load documents: %w", err)
+		}
+		for rows.Next() {
```

Comment:
I'm sure you're well aware of this -- I assume you're going for a "hello world" here -- but this is going to be extremely slow given the per-row back-and-forth.

I'd definitely recommend using the [Recovery Log with Idempotent Apply](https://docs.estuary.dev/reference/Connectors/materialization-protocol/#recovery-log-with-idempotent-apply) workflow to make this transactional / exactly once and also much, much faster.

As a general sketch:
 * During the Load phase the connector stages files to S3 of keys to load (example from [materialize-databricks ](https://github.com/estuary/connectors/blob/main/materialize-databricks/driver.go#L553)), and does either a single UNION'd query or per-binding sequential queries to yield all documents.
 * During the Store phase (e.x. [materialize-databricks](https://github.com/estuary/connectors/blob/main/materialize-databricks/driver.go#L668)) it similarly stages rows which should be COPY INTO'd or MERGE'd into the base table. But, they aren't actually applied yet. Instead, the connector just emits a transaction checkpoint state which enumerates idempotent queries that _should_ be run to apply the changes.
 * During the Acknowledge phase of the protocol, the connector runs (or re-runs) the idempotent queries from its current or recovered checkpoint state.
 
It's crucial that checkpoint'd are in fact idempotent. MERGE queries always are. COPY INTO can be with some finagling.
 
As another example, we've been moving our [materialize-snowflake](https://github.com/estuary/connectors/pull/1117) connector over to this same pattern, though that work is still in progress.

It doesn't look like Galaxy supports direct reads of JSON data from an external cloud storage table / stage. Are you thinking to have the connector write out temporary Hive tables for each transaction into a shared catalog, that then gets picked up and queried/merged against the base table?